app/Http/Controllers/Api/Web/StatsController.php

```diff
@@ -0,0 +1,159 @@
+<?php
+
+/**
+ * Pony.fm - A community for pony fan music.
+ * Copyright (C) 2015 Peter Deltchev
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+namespace Poniverse\Ponyfm\Http\Controllers\Api\Web;
+
+use Illuminate\Database\Eloquent\ModelNotFoundException;
+use Poniverse\Ponyfm\Http\Controllers\ApiControllerBase;
+use Poniverse\Ponyfm\Models\Track;
+use Auth;
+use DB;
+use Response;
+use Carbon\Carbon;
+
+class StatsController extends ApiControllerBase
+{
+    public function getTrackStatsHourly($id)
+    {
+        $query = DB::table('resource_log_items')
+            ->selectRaw('created_at AS time, COUNT(1) AS `plays`')
+            ->where('track_id', '=', $id)
+            ->where('log_type', '=', 3)
+            ->whereRaw('`created_at` > now() - INTERVAL 1 DAY')
+            ->groupBy('created_at')
+            ->get();
+
+        $now = Carbon::now();
+        $calcArray = array();
+        $output = array();
+
+        foreach($query as $item) {
+            $playDate = new Carbon($item->time);
+            $key = $playDate->diffInHours($now);
+            if (array_key_exists($key, $calcArray)) {
+                $calcArray[$key] += $item->plays;
+            } else {
+                $calcArray[$key] = $item->plays;
+            }
+        }
+
+        // Get the first key in the array (oldest play)
+        reset($calcArray);
+        $lastKey = (int) key($calcArray);
+
+        for ($i = 0; $i < $lastKey; $i++) {
+            if (!isset($calcArray[$i])) {
+                $calcArray[$i] = 0;
```

Comment:
I understand the purpose of this segment is to fill "missing" hours in the array with zeros. I propose an alternative approach: pre-filling `$calcArray` with 24 zeros (using `array_fill`), then incrementing the appropriate element for each hour in the first loop (above). You'll be able to eliminate this entire loop and the existence checks in the first one.